### PR TITLE
fix(add): add handling of merge option

### LINKED ISSUE #795: [BUG]: @webpack-cli/add not working on topScope and merge
**Describe the bug**
When we want to add something to `topScope` raises following error
Screenshot from 2019-03-18 20-10-56

**What is the current behavior?**
Currently this error raises as `add-generator` maps properties from `optionsSchema` of webpack. But this schema doesn't include properties that are part of `configuration` object we create in generator.

**To Reproduce**
1. `webpack add`
2. `give answer as topScope`

**Expected behavior**
Add items to topScope

**Screenshots**
uploaded above

**Please paste the results of `webpack-cli info` here, and mention other relevant information**
```

  System:
    OS: Linux 4.15 Ubuntu 18.04.2 LTS (Bionic Beaver)
    CPU: (3) x64 Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz
  Binaries:
    Node: 10.15.0 - ~/.nvm/versions/node/v10.15.0/bin/node
    Yarn: 1.13.0 - ~/.nvm/versions/node/v10.15.0/bin/yarn
    npm: 6.4.1 - ~/.nvm/versions/node/v10.15.0/bin/npm
  Browsers:
    Firefox: 65.0.1
  npmGlobalPackages:
    webpack-cli: 3.2.3
    webpack: 4.29.3
```
**Additional context**

### PR DESCRIPTION
**Pull request by rishabh3112**
BugFix #795 
Closed #795 
**What kind of change does this PR introduce?**

bugfix
**Did you add tests for your changes?**
N/A
**If relevant, did you update the documentation?**
N/A
**Summary**

**Does this PR introduce a breaking change?**

**Other information**

### CODE REVIEW

File: packages/generators/add-generator.ts [hunk 1]
```diff
@@ -173,6 +173,36 @@ export default class AddGenerator extends Generator {
...
+									} else {
+										return "Path doesn't corresponds to a javascript file";
+									}
+								}
+								return "Invalid path provided";
+							};
+							return this.prompt([
+								InputValidate(
+									"mergeFile",
+									"What is the location of webpack configuration file with which you want to merge current configuration?",
```
Comment (ematipico) [inline packages/generators/add-generator.ts]: Let's start doing something nice: could you move this long question inside a `constants` file?
Also, the same for the rest of the strings.

Ideally these strings could be exported from `webpack-cli/utils` so they can be reused across the board

File: packages/generators/add-generator.ts [hunk 2]
```diff
@@ -94,7 +93,8 @@ export default class AddGenerator extends Generator {
 		config: {
 			configName?: string,
 			topScope?: string[],
-			item?: string;
+			item?: string,
+			merge?: object,
```
Comment (ematipico) [inline packages/generators/add-generator.ts]: `object` is way too generic. Is there a proper type for this `merge` option?

**Comment by rishabh3112** [comment]
Hi @ematipico, I am pushing a sample refactoring ( will break lint as add-geneartor is importing non npm-published changes at `@webpack-cli/utils`.
If this refactoring seems fine to you, I will continue refactoring this and all generators as time permits in seperate PRs. :+1:

File: packages/generators/add-generator.ts (hunk 2 above)
Comment (rishabh3112) [reply packages/generators/add-generator.ts]: Added corresponding type @ematipico :+1:

**Comment by rishabh3112** [comment]
@ematipico Have a review when free 😅.

**Comment by rishabh3112** [comment]
CLA bot is showing that I have not signed 🤔. Some issue here @ematipico

**Comment by dhruvdutt** [comment]
@rishabh3112 It shows signed for me. All checks success.

File: packages/utils/generators/add/questions/index.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+export const topScopeQuestion = Input(
+	"topScope",
+	"What do you want to add to topScope?",
+);
+
+const mergeFileQuestionFunction = () => {
+ const question = "What is the location of webpack configuration with which you want to merge current configuration?";
+ const validator = (path: string) => {
+		const resolvedPath = resolve(process.env.PWD, path);
```
Comment (evenstensberg) [inline packages/utils/generators/add/questions/index.ts]: `process.cwd()`

File: packages/utils/generators/add/questions/index.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		if (existsSync(resolvedPath)) {
+			if (/\.js$/.test(path)) {
+				if (typeof require(resolvedPath) !== "object") {
+					return "Given file doesn't export an Object";
+				}
+				return true;
+			} else {
```
Comment (evenstensberg) [inline packages/utils/generators/add/questions/index.ts]: premature if-else

File: packages/generators/add-generator.ts [hunk 5]
```diff
@@ -162,16 +103,24 @@ export default class AddGenerator extends Generator {
...
 								this.configuration.config.topScope.push(topScopeAnswer.topScope);
 								done();
 							});
 						}
+						if (action === "merge") {
+							return this.prompt([mergeFileQuestion])
+							.then((mergeFileAnswer: {
+								mergeFile: string;
+							}) => {
+								const resolvedPath = resolve(process.env.PWD, mergeFileAnswer.mergeFile);
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: `process.cwd()`

File: packages/generators/add-generator.ts [hunk 6]
```diff
@@ -93,7 +47,8 @@ export default class AddGenerator extends Generator {
 		config: {
 			configName?: string,
 			topScope?: string[],
-			item?: string;
+			item?: string,
+			merge?: IWebpackOptions,
 			webpackOptions?: IWebpackOptions,
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: merge is in webpackOptions, feel free to make it a config prop but then the ast checking for merge needs to be removed. This is nice if we want to spit out a pure webpack config in webpackOptions

Comment (rishabh3112) [reply packages/generators/add-generator.ts]: @evenstensberg I also like to emit a pure config in final configuration built,
But with `merge` prop in `config` (like in current scenario of this PR), final configuration is still having merge prop in `module.exports`. Is that has to be changed too?

Comment (evenstensberg) [reply packages/generators/add-generator.ts]: No merge in module exports in my point being. This is ok but ast needs to be checked for code involving this

Comment (rishabh3112) [reply packages/generators/add-generator.ts]: The config has `merge` as prop only and is not in `webpackOptions` as you said [here](https://github.com/webpack/webpack-cli/blob/master/packages/utils/modify-config-helper.ts#discussion_r268748106)!
and `runTransform()` also consider `merge` as prop in config prior to my changes too.
https://github.com/webpack/webpack-cli/blob/b96857f182481623db2c1fc799fdab59dab81107/packages/utils/modify-config-helper.ts#L11-L19

So, IMO problem is solved as standalone AST parsing checking of merge is insignificant as it may be using functions outside the scope of object, thus parsing of final file makes more sense.

File: packages/utils/package.json [hunk 7]
```diff
@@ -31,6 +31,7 @@
     "@types/node": "^10.12.9",
     "@types/p-each-series": "^1.0.0",
     "@types/prettier": "^1.15.0",
+    "@webpack-cli/webpack-scaffold": "^0.1.5",
```
[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/utils/package.json]: Emanuele suggested moving questions to `@webpack-cli/utils`. And questions require `@webpack-cli/webpack-scaffold`, For methods like: `Confirm`, `List`, etc. So, that's why I have put that in dependencies

File: packages/utils/scaffold.ts [hunk 8]
```diff
@@ -48,7 +48,7 @@ export default function runTransform(transformConfig: ITransformConfig, action:
 			transformations.push("topScope");
 		}
 
-		if (config.merge) {
+		if (config.merge && transformations.indexOf("merge") === -1) {
```
[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/utils/scaffold.ts]: Here we don't want to add merge to transformations if it is already there. And, `indexOf("merge")` method returns `-1` when it is not present

[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/utils/scaffold.ts]: `indexOf("merge") >= 0` : Is exactly opposite of what this condition should do. It is checking if the merge is there in transformations. Will result in readdition of `merge` which is not expected
`~!indexOf("merge")` will also not work as anything other than `0` is true. thus wrong condition.

[the remark this answers is not in the document]
Comment (evenstensberg) [reply packages/utils/scaffold.ts]: Ok thanks for clearing out your intention. You could do `~!ndexOf("merge")`, which is a bitwise NOT to flip the bits. (Which is true for this case ;) )

File: packages/utils/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (evenstensberg) [reply packages/utils/package.json]: Nice, okay. I don't think user questions should be moved to `utils`, only src that touches functionality. If your functions using inquirer could be used in a general basis, don't put this in utils, but rather as a utils folder within the generator itself. This is to provide a seperation of concern / abstraction

File: packages/utils/scaffold.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/utils/scaffold.ts]: For this to work, order should be reversed IMO. i.e.
`!(~indexOf("merge"))`

[the remark this answers is not in the document]
Comment (evenstensberg) [reply packages/utils/scaffold.ts]: Yes. Anyway, I think it's fine, it's a nitpick

File: packages/utils/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/utils/package.json]: Okay I will do that way. Thanks :)

[the remark this answers is not in the document]
Comment (evenstensberg) [reply packages/utils/package.json]: I don't see it being used, so you could remove it

[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/utils/package.json]: Well currently is in use at `packages/utils/generators/add/questions/index.ts`
nevertheless, I removing all of the related code from `@webpack-cli/utils` and refactoring generator questions under `@webpack-cli/generators`. Hope that will be fine :)

**Comment by rishabh3112** [comment]
Done @evenstensberg.

File: packages/generators/add-generator.ts [hunk 9]
```diff
@@ -151,13 +104,29 @@ export default class AddGenerator extends Generator {
...
 									done();
 								}
 							);
 						}
+						if (action === "merge") {
+							return this.prompt([mergeFileQuestion])
+							.then((mergeFileAnswer: {
+								mergeFile: string;
+							}) => {
+								const resolvedPath = resolve(process.cwd(), mergeFileAnswer.mergeFile);
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: You'll need to parse this yourself. Extract this logic into a function `parseMergeConfig`, where you read the config, the `webpackOptions` is the module export of the config itself, you already have the filename from `basename(resolvedPath)` . I don't think that merge actually works like this. Merge takes a config name and attaches the name to a function. 

What you will need to do, is to make merge take in either `<Array>[name, path]` and add a require on top if the require is not in topScope (which I think is the best solution) or just add the name (bad DX).

So what you'll need to do, is to have this in the object of the generator:

```js
{
merge: ['webpackBaseConfig', 'path/to/foo'],
config {...}
}
```

and make the function parseMerge ( https://github.com/webpack/webpack-cli/blob/master/packages/utils/ast-utils.ts#L608-L640 ) accept an array that will convert the first idx to the function parameter, the second one to a topScope property. You could use the existing API to make this happen more easily, and that is to make a require statement based on the indices and add that to topScope before running the transform. This will leave you with the same API as it is, but providing support for `merge` in add.

File: packages/generators/utils/add/index.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
+import PROP_TYPES from "@webpack-cli/utils/prop-types";
```
Comment (evenstensberg) [inline packages/generators/utils/add/index.ts]: Other improvements is nice, cause CLI in add throws if dev-server is not a dependency, so you can keep the rest of the changes

File: packages/generators/add-generator.ts [hunk 11]
```diff
@@ -85,6 +42,12 @@ export default class AddGenerator extends Generator {
 			configName?: string;
 			topScope?: string[];
 			item?: string;
+			merge?: {
```
[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/generators/add-generator.ts]: Sure!
I had a problem with this one 😅, as the merge strictly can't have merge property, it will act as a bad interface. Can you suggest a workaround?

**Comment by evenstensberg** [comment]
Made a repro for you @rishabh3112 to help you get where I'm going with this:

https://gist.github.com/evenstensberg/87978d3e21bcc325f5fcef67f372fd68

File: packages/generators/add-generator.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (evenstensberg) [reply packages/generators/add-generator.ts]: I think maybe we should discuss how we should handle merge (see the comment on add generator) and then resume this if necessary

[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/generators/add-generator.ts]: Never mind, I just read the next comment, not a problem now :)

File: packages/generators/add-generator.ts (hunk 9 above)
Comment (rishabh3112) [reply packages/generators/add-generator.ts]: Nice solution.
Or we can just take up the path to the config file from user. Internally, add `webpack-merge` as dependency for the user and use it to scaffold a config that will use `webpack-merge`.
WDYT?

File: packages/generators/utils/add/index.ts (hunk 10 above)
Comment (rishabh3112) [reply packages/generators/utils/add/index.ts]: I didn't get what you want to say here?

File: packages/generators/add-generator.ts (hunk 9 above)
Comment (rishabh3112) [reply packages/generators/add-generator.ts]: Done @evenstensberg

File: packages/generators/add-generator.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/generators/add-generator.ts]: Solved now, Thanks!

File: packages/generators/utils/add/index.ts (hunk 10 above)
Comment (evenstensberg) [reply packages/generators/utils/add/index.ts]: Things that has been changed in logic not related to the transformation w/AST is fine

Comment (evenstensberg) [reply packages/generators/utils/add/index.ts]: Things that has been changed in logic not related to the transformation w/AST is fine

File: packages/utils/scaffold.ts [hunk 12]
```diff
@@ -11,6 +11,29 @@ import astTransform from "./recursive-parser";
 import runPrettier from "./run-prettier";
 import { Node } from "./types/NodePath";
 
+
+
+function mergeHandler(config: Config, transformations: string[]): [Config, string[]]{
+	if(!config["topScope"])
```
Comment (evenstensberg) [inline packages/utils/scaffold.ts]: Assuming you already declared the array in the constructor, you won't need to check if it's existent, you only need to push to the array

File: packages/utils/scaffold.ts [hunk 13]
```diff
@@ -45,16 +68,16 @@ export default function runTransform(transformConfig: TransformConfig, action: s
...
+			let transformations = mapOptionsToTransform(config);
 
 			if (config.topScope && transformations.indexOf("topScope") === -1) {
 				transformations.push("topScope");
 			}
 
-			if (config.merge) {
-				transformations.push("merge");
+			if (config.merge && transformations.indexOf("merge") === -1) {
+				[config, transformations] = mergeHandler(config, transformations);
```
Comment (evenstensberg) [inline packages/utils/scaffold.ts]: This is logic relevant to the generator providing the code, not the ast transform, this would make the abstraction bad. The generator is responsible to provide:

1. an array of indices with relevant data
2. The topScope property with the path

In the `handleMerge` function, you'll need to check if the part of the require is present in the most outerscope and append the second index if not present. 

Sorry for giving you an indication of what the easiest choice would be, it was a stepping stone towards a final solution

File: packages/utils/scaffold.ts [hunk 14]
```diff
@@ -45,16 +68,16 @@ export default function runTransform(transformConfig: TransformConfig, action: s
 
 	webpackConfig.forEach(
 		(scaffoldPiece: string): Promise<void> => {
-			const config: Config = transformConfig[scaffoldPiece];
+			let config: Config = transformConfig[scaffoldPiece];
```
Comment (evenstensberg) [inline packages/utils/scaffold.ts]: Config should be immutable/const here as we don't want to modify a presupplied config

File: packages/utils/scaffold.ts [hunk 15]
```diff
@@ -11,6 +11,29 @@ import astTransform from "./recursive-parser";
 import runPrettier from "./run-prettier";
 import { Node } from "./types/NodePath";
 
+
+
+function mergeHandler(config: Config, transformations: string[]): [Config, string[]]{
```
Comment (evenstensberg) [inline packages/utils/scaffold.ts]: I'm not a fan of this function as it isn't descriptive enough and it passes the transformation object in an indrection we don't want the AST to use. The transformation and it's data should have a 1:1 mapping

File: packages/utils/modify-config-helper.ts [hunk 16]
```diff
@@ -14,7 +14,7 @@ export interface Config extends Object {
 	};
 	topScope?: string[];
 	configName?: string;
-	merge: object;
+	merge: string|string[];
```
Comment (evenstensberg) [inline packages/utils/modify-config-helper.ts]: nice

File: packages/generators/utils/add/index.ts (hunk 10 above)
Comment (evenstensberg) [reply packages/generators/utils/add/index.ts]: It's all good is what I'm saying. We need to look at the merge logic first in the ast part and then we can look at this, but it looks ok

File: packages/generators/add-generator.ts (hunk 9 above)
Comment (evenstensberg) [reply packages/generators/add-generator.ts]: Nice. Have a look at the comment further down

File: packages/utils/scaffold.ts (hunk 13 above)
Comment (evenstensberg) [reply packages/utils/scaffold.ts]: To give you a hint: We're pushing the transformation, which means that it's probably the latest transform to be run, so what you'll need to do, is this logic, but directly as an ast transform. The data gets passed down by the each astTransform func

Comment (rishabh3112) [reply packages/utils/scaffold.ts]: Thanks for review! As far as I could understand now, I need to update `parseMerge` function directly in `ast-utills` 
to just use the 1st index of the array 
And push to the topScope here itself. Right?

File: packages/utils/scaffold.ts (hunk 12 above)
Comment (rishabh3112) [reply packages/utils/scaffold.ts]: Added to avoid errors if runTransform function is directly used, like I do in webpack UI

File: packages/utils/scaffold.ts (hunk 14 above)
Comment (rishabh3112) [reply packages/utils/scaffold.ts]: I was updating config later using `mergeHandler`, that's why this change. See the comment above for more info on my view on this now.

File: packages/utils/scaffold.ts (hunk 15 above)
Comment (rishabh3112) [reply packages/utils/scaffold.ts]: Remove this function

File: packages/utils/scaffold.ts (hunk 14 above)
Comment (rishabh3112) [reply packages/utils/scaffold.ts]: Updated

File: packages/utils/scaffold.ts (hunk 12 above)
Comment (rishabh3112) [reply packages/utils/scaffold.ts]: Will found some other way around for UI 👍 . Removed from here.

**Comment by rishabh3112** [comment]
Done @evenstensberg

File: packages/generators/add-generator.ts [hunk 17]
```diff
@@ -151,13 +99,25 @@ export default class AddGenerator extends Generator {
...
 						}
+						if (action === "merge") {
+							return this.prompt(mergeFileQuestion)
+							.then((mergeFileAnswer: {
+								mergeFile: string,
+								mergeConfigName: string
+							}) => {
+								const resolvedPath = resolve(process.cwd(), mergeFileAnswer.mergeFile);
+								// eslint-disable-next-line
+								this.configuration.config["merge"] = [mergeFileAnswer.mergeConfigName, resolvedPath];
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: `action` here instead of hard-coded value

File: packages/utils/recursive-parser.ts [hunk 18]
```diff
@@ -15,7 +15,7 @@ export default function recursiveTransform(
 		console.error("Error in parsing top scope, Array required");
 		return false;
 	} else if (key === "merge") {
-		if (typeof value === "string") {
+		if (typeof value[0] === "string" && Array.isArray(value)) {
```
Comment (evenstensberg) [inline packages/utils/recursive-parser.ts]: both indices are strings here, start by checking array, then that both indices are strings

File: packages/utils/scaffold.ts [hunk 19]
```diff
@@ -53,10 +54,17 @@ export default function runTransform(transformConfig: TransformConfig, action: s
 				transformations.push("topScope");
 			}
 
-			if (config.merge) {
+			if (config.merge && transformations.indexOf("merge") === -1) {
```
Comment (evenstensberg) [inline packages/utils/scaffold.ts]: This is a double up if, you can nest into a single one

File: packages/utils/scaffold.ts [hunk 20]
```diff
@@ -53,10 +54,17 @@ export default function runTransform(transformConfig: TransformConfig, action: s
...
 				transformations.push("merge");
 			}
 
+			if (config.merge) {
```
Comment (evenstensberg) [inline packages/utils/scaffold.ts]: nest this ^

File: packages/utils/ast-utils.ts [hunk 21]
```diff
@@ -605,7 +605,7 @@ function parseTopScope(j: JSCodeshift, ast: Node, value: string[], action: strin
  */
 
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
-function parseMerge(j: JSCodeshift, ast: Node, value: string, action: string): boolean | Node {
+function parseMerge(j: JSCodeshift, ast: Node, value: string[], action: string): boolean | Node {
 	function createMergeProperty(p: Node): boolean {
 		// FIXME Use j.callExp()
```
Comment (evenstensberg) [inline packages/utils/ast-utils.ts]: Split into two variables, give them meaningful names. First one is supposed to go here, and then you'll need to check if the require of this is present in the outer body of the AST. If it's present, you don't need to do anything, if it's not existent you will need to add it. 

Remove the topScope push and manually add that here. You could reuse the logic used in topScope to a certain extent. Remember to abstract this code so it looks neat

File: packages/utils/ast-utils.ts [hunk 22]
```diff
@@ -626,14 +626,35 @@ function parseMerge(j: JSCodeshift, ast: Node, value: string, action: string): b
...
 		};
 		(p.value as Node).body[bodyLength - 1] = newVal;
 		return false; // TODO: debug later
 	}
+
+	function addMergeImports(ast: Node, mergeImports: string[]) {
+		ast.find(j.Program).filter((program: Node): boolean => {
+			mergeImports.forEach(
+				(imp: string): void => {
+					if ((program.value as Node).body.indexOf(imp) === -1){
```
Comment (rishabh3112) [inline packages/utils/ast-utils.ts]: Currently checking whether in the current iteration these two imports are added or not. Similar to `parseTopScope` which also not check this.

I want to make a better checking if a particular package is already imported or not by iterating over all `CallExpression`s and checking values of `argument` where `callee` is `require`. This same function could also be used for `parseTopScope` for better checking when imports are pushed to the topScope.
Should I do that in this PR itself on open a new one for this? asking as scope of that change is larger that that of this PR.

**Comment by rishabh3112** [comment]
Rebased, look at the above review.

File: packages/utils/ast-utils.ts (hunk 22 above)
Comment (evenstensberg) [reply packages/utils/ast-utils.ts]: Take it in this PR, but could you rephrase your naming conventions such that it makes sense

File: packages/utils/scaffold.ts (hunk 13 above)
Comment (evenstensberg) [reply packages/utils/scaffold.ts]: aye

File: packages/utils/ast-utils.ts (hunk 22 above)
Comment (rishabh3112) [reply packages/utils/ast-utils.ts]: Like?

Comment (evenstensberg) [reply packages/utils/ast-utils.ts]: What is an `imp` ? what does it do?

File: packages/utils/ast-utils.ts [hunk 23]
```diff
@@ -626,14 +626,35 @@ function parseMerge(j: JSCodeshift, ast: Node, value: string, action: string): b
...
+	}
+
 	if (value) {
-		return ast.find(j.Program).filter((p: Node): boolean => createMergeProperty(p));
+		const [configIdentifier, configPath] = value;
+		const mergeImports = [
+			`const ${configIdentifier} = require('${configPath}')`,
+			`const merge = require('webpack-merge')`
+		];
+		ast = addMergeImports(ast, mergeImports);
```
Comment (evenstensberg) [inline packages/utils/ast-utils.ts]: Use the jscodeshift api for this together with `replace`

File: packages/utils/ast-utils.ts (hunk 22 above)
Comment (rishabh3112) [reply packages/utils/ast-utils.ts]: Short for `import`, import is a reserved keyword and can't use as a variable name.

It is one of the imports in `mergeImports` array.

Comment (evenstensberg) [reply packages/utils/ast-utils.ts]: exactly, just getting you on the right track here from a different perspective

Comment (rishabh3112) [reply packages/utils/ast-utils.ts]: Okay 😂, any other naming you felt bad ?

**Comment by rishabh3112** [comment]
Done @evenstensberg

**Comment by rishabh3112** [comment]
@evenstensberg can I know if anything left to be done on this? Also, I suggest you merge this first before 
#843 as that way I would be able to fix conflicts for obvious file `add-generator.ts` at #843.

**Comment by evenstensberg** [comment]
Going through it now

File: packages/utils/ast-utils.ts [hunk 24]
```diff
@@ -626,14 +637,27 @@ function parseMerge(j: JSCodeshift, ast: Node, value: string, action: string): b
...
 		(p.value as Node).body[bodyLength - 1] = newVal;
 		return false; // TODO: debug later
 	}
+
+	function addMergeImports(configIdentifier: string, configPath: string) {
+		ast.find(j.Program).forEach(p => {
+			if (!isImportPresent(j, ast, 'webpack-merge')) {
+				(p.value as Node).body.splice(-1, 0, `const merge = require('webpack-merge')`);
+			}
+			(p.value as Node).body.splice(-1, 0, `const ${configIdentifier} = require('${configPath}')`);
```
Comment (evenstensberg) [inline packages/utils/ast-utils.ts]: you need to check if this import is present too, look for the configIdentifier in the body. So a new `if` here: `if importPresent(configIdentifier)`

File: packages/utils/ast-utils.ts [hunk 25]
```diff
@@ -626,14 +637,27 @@ function parseMerge(j: JSCodeshift, ast: Node, value: string, action: string): b
...
+		})
+	}
+
 	if (value) {
-		return ast.find(j.Program).filter((p: Node): boolean => createMergeProperty(p));
+		const [configIdentifier, configPath] = value;
```
Comment (evenstensberg) [inline packages/utils/ast-utils.ts]: Gotta check these types if they're present

File: packages/utils/recursive-parser.ts [hunk 26]
```diff
@@ -15,7 +15,7 @@ export default function recursiveTransform(
 		console.error("Error in parsing top scope, Array required");
 		return false;
 	} else if (key === "merge") {
-		if (typeof value === "string") {
+		if (Array.isArray(value) && typeof value[0] === "string" && typeof value[1] === "string") {
```
Comment (evenstensberg) [inline packages/utils/recursive-parser.ts]: We can skip these comparrisons until later on in the import functions you've added. Checking for array is sufficient here, comparing the destructed values later

File: packages/utils/ast-utils.ts (hunk 25 above)
Comment (rishabh3112) [reply packages/utils/ast-utils.ts]: Throwing errors now in case of type mismatch.

Comment (evenstensberg) [reply packages/utils/ast-utils.ts]: awesome sauce

**Comment by rishabh3112** [comment]
Done @evenstensberg

**Comment by evenstensberg** [comment]
Mind rebasing @rishabh3112 ? I'll have a look, just got caught up in other issues

**Comment by evenstensberg** [comment]
Hmu once that's done and I'l do a review to get this merged

**Comment by rishabh3112** [comment]
Done @evenstensberg

File: packages/generators/add-generator.ts [hunk 27]
```diff
@@ -1,57 +1,25 @@
...
- *
- * @returns {Promise} Returns promise which resolves to filtered props
- *
- */
-const searchProps = (answers: object, input: string): Promise<string[]> => {
-	input = input || "";
-	return Promise.resolve(PROPS.filter((prop: string): boolean => prop.toLowerCase().includes(input.toLowerCase())));
-};
+import entryQuestions from "./utils/entry";
+import * as AutoComplete from "inquirer-autocomplete-prompt";
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: Move package installed deps as the first imports

File: packages/generators/add-generator.ts [hunk 28]
```diff
@@ -136,13 +94,23 @@ export default class AddGenerator extends Generator {
...
 									this.configuration.config.topScope.push(topScopeAnswer.topScope);
 									done();
 								}
 							);
 						}
+						if (action === "merge") {
+							return this.prompt(mergeFileQuestion).then(
+								(mergeFileAnswer: { mergeFile: string; mergeConfigName: string }): void => {
+									const resolvedPath = resolve(process.cwd(), mergeFileAnswer.mergeFile);
+									// eslint-disable-next-line
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: Don't disable this lint rule

File: packages/generators/add-generator.ts [hunk 29]
```diff
@@ -1,57 +1,25 @@
...
-import * as path from "path";
+import {
+	actionTypeQuestion,
+	entryTypeQuestion,
+	manualOrListInput,
+	mergeFileQuestion,
+	topScopeQuestion
+} from "./utils/add/questions";
+
+import { traverseAndGetProperties, webpackDevServerSchema, webpackSchema } from "./utils/add";
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: You don't need to abstract the schemas

File: packages/generators/utils/add/questions/index.ts [hunk 30]
```diff
@@ -0,0 +1,63 @@
+import {
```
Comment (evenstensberg) [inline packages/generators/utils/add/questions/index.ts]: rename to questions.js to avoid parent index and this index.js being confusing

File: packages/generators/utils/add/questions/index.ts [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+
+const mergeFileQuestionsFunction = (): Question[] => {
+ const mergePathQuestion = "What is the location of webpack configuration with which you want to merge current configuration?";
+ const mergePathValidator = (path: string): boolean|string => {
+		const resolvedPath = resolve(process.cwd(), path);
+		if (existsSync(resolvedPath)) {
+			if (/\.js$/.test(path)) {
+				return true;
+			}
+			return "Path doesn't corresponds to a javascript file";
```
Comment (evenstensberg) [inline packages/generators/utils/add/questions/index.ts]: doesn't correspond

File: packages/generators/utils/add/questions/index.ts [hunk 32]
```diff
@@ -0,0 +1,63 @@
+import {
+	AutoComplete,
```
Comment (evenstensberg) [inline packages/generators/utils/add/questions/index.ts]: Could you rephrase your questions to the ones that are on master right now?

File: packages/utils/ast-utils.ts [hunk 33]
```diff
@@ -1,6 +1,24 @@
 import { JSCodeshift, Node, valueType } from "./types/NodePath";
 import * as validateIdentifier from "./validate-identifier";
 
+function isImportPresent(j: JSCodeshift, ast: Node, path: string): boolean {
+	if (typeof path !== "string") {
+		throw new Error(`path parameter should be string, recieved ${typeof path}`);
+	}
+	let isPresent = false;
```
Comment (evenstensberg) [inline packages/utils/ast-utils.ts]: `mergeImportExists`

File: packages/utils/ast-utils.ts [hunk 34]
```diff
@@ -626,14 +644,37 @@ function parseMerge(j: JSCodeshift, ast: Node, value: string, action: string): b
...
 		};
+
 		(p.value as Node).body[bodyLength - 1] = newVal;
 		return false; // TODO: debug later
 	}
+
+	function addMergeImports(configIdentifier: string, configPath: string): void {
+		if (typeof configIdentifier !== "string" || typeof configPath !== "string") {
+			throw new Error(
+				`Both parameters should be string. recieved ${typeof configIdentifier}, ${typeof configPath}`
```
Comment (evenstensberg) [inline packages/utils/ast-utils.ts]: strings

File: packages/utils/ast-utils.ts (hunk 33 above)
Comment (rishabh3112) [reply packages/utils/ast-utils.ts]: It is a general utility function, can be used for `topScope` too while adding new packages. Are you okay with "importExists"?

File: packages/generators/utils/add/questions/index.ts (hunk 32 above)
Comment (rishabh3112) [reply packages/generators/utils/add/questions/index.ts]: Questions are the same as in the master

**Comment by rishabh3112** [comment]
Have applied suggestions @evenstensberg and left some comments for you :)

File: packages/generators/add-generator.ts [hunk 35]
```diff
@@ -1,57 +1,26 @@
...
-import * as autoComplete from "inquirer-autocomplete-prompt";
-import * as path from "path";
-
 import npmExists from "@webpack-cli/utils/npm-exists";
 import { getPackageManager } from "@webpack-cli/utils/package-manager";
-import PROP_TYPES from "@webpack-cli/utils/prop-types";
-import { AutoComplete, Confirm, Input, List } from "@webpack-cli/webpack-scaffold";
+import { Input, List } from "@webpack-cli/webpack-scaffold";
+// eslint-disable-next-line
+export const webpackDevServerSchema = require("webpack-dev-server/lib/options.json");
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: with ts you can import without require

File: packages/generators/add-generator.ts [hunk 36]
```diff
@@ -1,57 +1,26 @@
...
- *
- */
-const searchProps = (answers: object, input: string): Promise<string[]> => {
-	input = input || "";
-	return Promise.resolve(PROPS.filter((prop: string): boolean => prop.toLowerCase().includes(input.toLowerCase())));
-};
+import entryQuestions from "./utils/entry";
+import { generatePluginName } from "./utils/plugins";
+// eslint-disable-next-line
+export const webpackSchema = require("../optionsSchema.json");
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: same here

File: packages/generators/add-generator.ts (hunk 35 above)
Comment (rishabh3112) [reply packages/generators/add-generator.ts]: Done @evenstensberg

File: packages/generators/add-generator.ts [hunk 37]
```diff
@@ -1,57 +1,24 @@
...
-
 import npmExists from "@webpack-cli/utils/npm-exists";
 import { getPackageManager } from "@webpack-cli/utils/package-manager";
-import PROP_TYPES from "@webpack-cli/utils/prop-types";
-import { AutoComplete, Confirm, Input, List } from "@webpack-cli/webpack-scaffold";
+import { Input, List } from "@webpack-cli/webpack-scaffold";
+import webpackDevServerSchema from "webpack-dev-server/lib/options.json";
+import * as AutoComplete from "inquirer-autocomplete-prompt";
+import path, { resolve } from "path";
+import glob from "glob-all";
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: glob and path on top of import, then the autocomplete. Add a space after that and the cli imports after that

```
const fs ...
const path ....
const autocomplete

const cli..
const cli2..
```

File: packages/generators/utils/add/questions.ts [hunk 38]
```diff
@@ -0,0 +1,46 @@
+import { AutoComplete, Confirm, Input, InputValidate } from "@webpack-cli/webpack-scaffold";
+import { existsSync } from "fs";
```
Comment (evenstensberg) [inline packages/generators/utils/add/questions.ts]: same import order as mentioned in prev file

**Comment by rishabh3112** [comment]
Have moved local imports (example `../utils`)  at last.

File: packages/generators/add-generator.ts [hunk 39]
```diff
@@ -1,57 +1,25 @@
...
+import * as AutoComplete from "inquirer-autocomplete-prompt";
 import * as Generator from "yeoman-generator";
-
-import * as glob from "glob-all";
-import * as autoComplete from "inquirer-autocomplete-prompt";
-import * as path from "path";
+import webpackDevServerSchema from "webpack-dev-server/lib/options.json";
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: group this with the ones below

File: packages/generators/utils/add/questions.ts [hunk 40]
```diff
@@ -0,0 +1,48 @@
...
+import { resolve } from "path";
+import { Question } from "inquirer";
+
+import { AutoComplete, Confirm, Input, InputValidate } from "@webpack-cli/webpack-scaffold";
+
```
Comment (evenstensberg) [inline packages/generators/utils/add/questions.ts]: - whitespace, group these

**Comment by rishabh3112** [comment]
No problem, Thanks🎉🎉
